### PR TITLE
Employee

### PR DESCRIPTION
**Pull request by pushkarjoshi19152**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -0,0 +1,4 @@
+
```
Comment (rehannawab) [inline .gitignore:1]: application.properties should not be ignored. Remove this file from PR

File: Server/.gitignore [hunk 2]
```diff
@@ -28,6 +28,7 @@ target/
 build/
 !**/src/main/**/build/
 !**/src/test/**/build/
+.application.properties
```
Comment (rehannawab) [inline Server/.gitignore:31]: remove this

File: Server/pom.xml [hunk 3]
```diff
@@ -33,17 +33,18 @@
...
-			<groupId>com.mysql</groupId>
-			<artifactId>mysql-connector-j</artifactId>
-			<scope>runtime</scope>
-		</dependency>
 		<dependency>
 			<groupId>org.mariadb.jdbc</groupId>
 			<artifactId>mariadb-java-client</artifactId>
 			<version>2.7.3</version>
 		</dependency>
+		<dependency>
```
Comment (rehannawab) [inline Server/pom.xml:41]: remove this

File: Server/src/main/java/com/abs/errp/achievement/AchievementServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+	private AchievementRepository achievementRepository;
+
+	@Autowired
+	LoggedInUserContext userContext;
+
+	// Check if user is authorized for the action
+	// achievement's employeeId should be same as logged in user's id
+
+	public boolean isAuthorized(Achievement achievement) {
+		LoggedInUser user = this.userContext.getLoggedInUser();
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/achievement/AchievementServiceImpl.java:30]: Creating or updating achievement should add the userId on the server instead of sending this from UI

File: Server/src/main/java/com/abs/errp/achievement/AchievmentController.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	@GetMapping("/paginated/{pageNo}/{pageSize}")
+	public List<Achievement> getPaginatedCountries(@PathVariable int pageNo,
+			@PathVariable int pageSize) {
+
+		return achievementService.findPaginated(pageNo, pageSize);
+	}
+
+	@PostMapping("/update/{id}")
+	public ResponseEntity<Achievement> updateAchievement(@PathVariable int id, @RequestBody Achievement achievement) {
+		System.out.println("update id = " + id);
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/achievement/AchievmentController.java:41]: remove sysout

File: Server/src/main/java/com/abs/errp/achievement/AchievmentController.java [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+public class AchievmentController {
+
+	private AchievementService achievementService;
+
+	public AchievmentController(AchievementService achievementService) {
+		super();
+		this.achievementService = achievementService;
+	}
+
+	@PostMapping("/add")
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/achievement/AchievmentController.java:26]: controller mapping should be on the class level. you should only add empty PUT/POST/GET requests in the controller. For getting id only map id in the reqeust param.

File: Server/src/main/java/com/abs/errp/awardrequest/AwardRequestServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+	@Override
+	public List<AwardRequest> getAllAwardRequest() {
+
+		return awardServiceRepository.findAll();
+	}
+
+	@Override
+	public AwardRequest saveAwardRequest(AwardRequest awardRequest) {
+
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/awardrequest/AwardRequestServiceImpl.java:25]: should there be validations while saving award request?

File: Server/src/main/java/com/abs/errp/awardtype/AwardTypeController.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+public class AwardTypeController {
+
+	private AwardTypeService awardTypeService;
+
+	public AwardTypeController(AwardTypeService awardTypeService) {
+		super();
+		this.awardTypeService = awardTypeService;
+	}
+
+	@GetMapping("/awardtypes")
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/awardtype/AwardTypeController.java:23]: the mapping should be on the controller level

File: Server/src/main/java/com/abs/errp/awardtype/AwardTypeController.java [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+	public List<AwardType> getAllAwardTypes()
+
+	{
+		return awardTypeService.getAllAwardTypes();
+	}
+
+	@GetMapping("/users")
+	public List<ErrpUser> getAllUsers() {
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/awardtype/AwardTypeController.java:31]: why is this endpoint in this controller? and why is this needed? for large employee result set this will kill the UI

File: Server/src/main/java/com/abs/errp/awardtype/AwardTypeServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,38 @@
...
+		this.awardServiceRespository = awardServiceRespository;
+	}
+
+	@Override
+	public List<AwardType> getAllAwardTypes() {
+		return awardServiceRespository.findAll();
+	}
+
+	@Override
+	public List<ErrpUser> getAllUsers() {
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/awardtype/AwardTypeServiceImpl.java:34]: this should not be here

File: Server/src/main/java/com/abs/errp/awardtype/ErrpUserRepository.java [hunk 11]
```diff
@@ -0,0 +1,9 @@
+package com.abs.errp.awardtype;
+
+import org.springframework.data.jpa.repository.JpaRepository;
+
+import com.abs.errp.user.ErrpUser;
+
+public interface ErrpUserRepository extends JpaRepository<ErrpUser, Integer> {
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/awardtype/ErrpUserRepository.java:7]: there is already a repository for ErrpUser correct?

File: Server/src/main/java/com/abs/errp/employeeaward/EmployeeAwardController.java [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+		super();
+		this.employeeawardService = employeeawardService;
+	}
+
+	@GetMapping()
+	public List<EmployeeAward> getAllEmployeeAward()
+
+	{
+
+		return employeeawardService.getAllEmployeeAward();
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/employeeaward/EmployeeAwardController.java:27]: why is this needed?

File: Server/src/main/java/com/abs/errp/exception/ErrpExceptionHandler.java [hunk 13]
```diff
@@ -15,16 +15,15 @@
...
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
 @ControllerAdvice
-public class ErrpExceptionHandler {
+public class ErrpExceptionHandler extends RuntimeException {
+
+	/**
+	 * 
+	 */
+	private static final long serialVersionUID = 1L;
+
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/exception/ErrpExceptionHandler.java:27]: why was this file changed?

File: Server/src/main/java/com/abs/errp/security/LoginController.java [hunk 14]
```diff
@@ -15,13 +15,16 @@
...
+import org.springframework.web.bind.annotation.CrossOrigin;
 import org.springframework.web.bind.annotation.GetMapping;
 import org.springframework.web.bind.annotation.RequestBody;
 import org.springframework.web.bind.annotation.RequestMapping;
 import org.springframework.web.bind.annotation.RequestMethod;
 import org.springframework.web.bind.annotation.ResponseBody;
 import org.springframework.web.bind.annotation.RestController;
 
+import com.abs.errp.user.ErrpUser;
+
```
Comment (rehannawab) [inline Server/src/main/java/com/abs/errp/security/LoginController.java:27]: revert changes to this file

File: Server/src/main/resources/application.properties [hunk 15]
```diff
@@ -1,7 +1,23 @@
 spring.datasource.driver-class-name=org.mariadb.jdbc.Driver
-spring.datasource.url=jdbc:mariadb://localhost:3306/errp
-spring.datasource.username=
-spring.datasource.password=
-spring.jpa.show-sql=true
+spring.datasource.url=jdbc:mariadb://localhost:3307/errp
```
Comment (rehannawab) [inline Server/src/main/resources/application.properties:2]: revert checkin to this file

File: UI/angular.json [hunk 16]
```diff
@@ -136,7 +136,8 @@
   "cli": {
     "schematicCollections": [
       "@ionic/angular-toolkit"
-    ]
+    ],
```
Comment (rehannawab) [inline UI/angular.json:139]: revert this checkin

File: UI/src/app/Achievement/achievement-form/achievement-form.component.html [hunk 17]
```diff
@@ -0,0 +1,53 @@
...
+            <ion-back-button></ion-back-button>
+        </ion-buttons>
+        <ion-title>{{ isIdPresent ? 'Update Achievement' : 'Add Achievement' }}</ion-title>
+    </ion-toolbar>
+</ion-header>
+<ion-content class="ion-padding">
+    <ion-content>
+        <ion-card>
+            <ion-card-content>
+                <ion-card-header>
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-form/achievement-form.component.html:13]: remove unnecessary header

File: UI/src/app/Achievement/achievement-form/achievement-form.component.scss [hunk 18]
```diff
@@ -0,0 +1,5 @@
+ion-card-header.ios {
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-form/achievement-form.component.scss:1]: why this css is needed?

File: UI/src/app/Achievement/achievement-form/achievement-form.component.ts [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+
+    handleUpdate() {
+        this.updateAchievement();
+    }
+
+    handleSubmit() {
+        this.submitAchievement();
+    }
+
+    submitAchievement() {
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-form/achievement-form.component.ts:80]: lots of common code in submit/create/update achievement. Put common code in one method, dont repeat same code multiple times

File: UI/src/app/Achievement/achievement-form/achievement-form.component.ts [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+        const today = new Date();
+        this.achievement.createdBy = this.loggedInUser?.employeeId;
+        this.achievement.updatedBy = this.loggedInUser?.employeeId;
+        this.achievement.createdDate = today;
+        this.achievement.updatedDate = today;
+        this.achievement.employeeId = this.loggedInUser?.employeeId;
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-form/achievement-form.component.ts:86]: this should be updated on the server

File: UI/src/app/Achievement/achievement-list/achievement-list.component.html [hunk 21]
```diff
@@ -0,0 +1,60 @@
...
+    </ion-toolbar>
+</ion-header>
+<ion-content>
+    <ion-card *ngFor="let achievement of achievements">
+        <ion-card-header>
+            <ion-card-subtitle class="ion-text-right">
+                {{ achievement.updatedDate | amDateFormat : 'LL' }}
+            </ion-card-subtitle>
+            <ion-card-title>{{ achievement.title }}</ion-card-title>
+            <ion-card-title *ngIf="achievement.achievementStatus == 0" class="ion-text-right"> Saved </ion-card-title>
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-list/achievement-list.component.html:16]: 'Saved' should not be displayed

File: UI/src/app/Achievement/achievement-list/achievement-list.component.html [hunk 22]
```diff
@@ -0,0 +1,60 @@
...
+            <ion-card-title *ngIf="achievement.achievementStatus == 1" class="ion-text-right"> Pending </ion-card-title>
+            <ion-card-title *ngIf="achievement.achievementStatus == 2" class="ion-text-right">approved</ion-card-title>
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-list/achievement-list.component.html:18]: 'A' should be capital

File: UI/src/app/Achievement/achievement-list/achievement-list.component.html [hunk 23]
```diff
@@ -0,0 +1,60 @@
...
+
+        <ion-grid>
+            <ion-row>
+                <ion-col></ion-col>
+                <ion-col size="auto"
+                    ><ion-button
+                        fill="solid"
+                        *ngIf="achievement.achievementStatus == 0"
+                        routerLink="editAchievement/{{ achievement.achievementId }}"
+                        >Update<ion-icon slot="end" name="create"></ion-icon></ion-button
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-list/achievement-list.component.html:32]: use font awesome icons

File: UI/src/app/Achievement/achievement-list/achievement-list.component.ts [hunk 24]
```diff
@@ -0,0 +1,137 @@
...
+      for (let i = 0; i < data.length; i++) {
+        this.achievements.push(data[i]);
+      }
+      if (isFirstLoad) {
+        event.target.complete();
+      }
+      this.pageNo++;
+
+    }, error => {
+      console.error(error);
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-list/achievement-list.component.ts:54]: add proper error handling

File: UI/src/app/Achievement/achievement-list/achievement-list.component.ts [hunk 25]
```diff
@@ -0,0 +1,137 @@
...
+    })
+
+  }
+  submitAchievement(achievement: Achievement) {
```
Comment (rehannawab) [inline UI/src/app/Achievement/achievement-list/achievement-list.component.ts:58]: if there is common code in submit achievement from list and form, put that common code in the service class

File: UI/src/app/Awards/award-form/award-form.component.ts [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+
+  // }]
+  awards:AwardType[];
+
+
+
+
+  ngOnInit() {
+    this.awardService.getAwardTypes().subscribe(data=>{
+      this.awards = data;
```
Comment (rehannawab) [inline UI/src/app/Awards/award-form/award-form.component.ts:40]: add error handling

**Review by rehannawab — CHANGES_REQUESTED** [review]

File: Server/src/main/java/com/abs/errp/DatasourceConfig.java [hunk 27]
```diff
@@ -7,14 +7,14 @@
 import org.springframework.context.annotation.Configuration;
 
 //@Configuration
-public class DatasourceConfig {
```
Comment (rbaidABS) [inline Server/src/main/java/com/abs/errp/DatasourceConfig.java:10]: Keep a check while committing files. This is common file which does not need to be checkedin

File: Server/src/main/java/com/abs/errp/awardrequest/AwardRequestServiceImpl.java (hunk 7 above)
Comment (pushkarjoshi19152) [reply Server/src/main/java/com/abs/errp/awardrequest/AwardRequestServiceImpl.java:25]: This is from raise an award story we are yet to complete it

File: Server/src/main/java/com/abs/errp/achievement/AchievmentController.java [hunk 28]
```diff
@@ -0,0 +1,64 @@
...
+			@PathVariable int pageSize) {
+
+		return achievementService.findPaginated(pageNo, pageSize);
+	}
+
+	@PostMapping("/update/{id}")
+	public ResponseEntity<Achievement> updateAchievement(@PathVariable int id, @RequestBody Achievement achievement) {
+		System.out.println("update id = " + id);
+		return new ResponseEntity<Achievement>(achievementService.updateAchievement(id, achievement),
+				HttpStatus.CREATED);
```
Comment (rbaidABS) [inline Server/src/main/java/com/abs/errp/achievement/AchievmentController.java:43]: After successful update, the response code should be HttpStatus.OK (200) or HttpStatus.NO_CONTENT (204). Here the status code being used for Update is incorrect

File: Server/src/main/java/com/abs/errp/awardrequest/AwardRequestController.java [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+		super();
+		System.out.println("2");
+		this.awardRequestService = awardRequestService;
+	}
+
+	@GetMapping()
+	public List<AwardRequest> getAllAwardRequest()
+
+	{
+		System.out.println("AwardRequest");
```
Comment (rbaidABS) [inline Server/src/main/java/com/abs/errp/awardrequest/AwardRequestController.java:31]: Self code review required before submitting the PR. Remove Sysouts from all the places

Comment (pushkarjoshi19152) [reply Server/src/main/java/com/abs/errp/awardrequest/AwardRequestController.java:31]: This is from raise an award story we are still working on it

**Review by rehannawab — APPROVED** [review]
